bioio_ome_zarr/writers/ome_zarr_writer_v2.py

```diff
@@ -75,84 +75,130 @@ class OMEZarrWriter:
-                chunks=level_chunk_sizes[i],
-                dtype=dtype,
-                compressor=compressor,
-                zarr_format=2,
+        if overwrite:
+            # OVERWRITE MODE: create all new levels
+            self.root = zarr.group(
+                store=self.store,
+                overwrite=True,
+                zarr_version=2,
+            )
+            self._existing_level_count = 0
+            self.levels = []
+            for idx, (shape, chunk) in enumerate(zip(shapes, chunk_sizes)):
+                self._create_level(idx, shape, chunk, dtype, compressor)
+            return
+        else:
+            # APPEND MODE: detect which levels already exist
+            self.root = zarr.open_group(
+                store=self.store,
+                mode="a",
+                zarr_version=2,
             )
-            self.levels.append(ZarrLevel(shape, level_chunk_sizes[i], dtype, arr))
+            try:
+                existing_idxs = list(Reader(self.output_path).resolution_levels)
+            except Exception:
+                existing_idxs = []
+            self._existing_level_count = len(existing_idxs)
+
+            # load existing levels
+            self.levels = []
+            for idx in existing_idxs:
+                arr = zarr.open_array(store=self.store, path=str(idx), mode="a")
+                shape = shapes[idx]
+                chunk = chunk_sizes[idx]
+                self.levels.append(ZarrLevel(shape, chunk, arr.dtype, arr))
+
+            # create only the new levels beyond existing_idxs
+            for new_idx in range(self._existing_level_count, len(shapes)):
+                shape = shapes[new_idx]
+                chunk = chunk_sizes[new_idx]
+                self._create_level(new_idx, shape, chunk, dtype, compressor)
 
     def _downsample_and_write_batch_t(
         self, data_tczyx: da.Array, start_t: int, end_t: int, toffset: int = 0
     ) -> None:
+        """
+        Write frames [start_t, end_t) into each new level only.
+        Levels with index < self._existing_level_count are skipped.
+        """
         dtype = data_tczyx.dtype
-        for k in range(start_t, end_t):
-            subset = data_tczyx[[k - start_t]]
-            da.to_zarr(
-                subset,
-                self.levels[0].zarray,
-                region=(slice(k + toffset, k + toffset + 1),),
-            )
-        for j in range(1, len(self.levels)):
-            nextshape = (end_t - start_t,) + self.levels[j].shape[1:]
-            data_tczyx = resize(data_tczyx, nextshape, order=0).astype(dtype)
-            for k in range(start_t, end_t):
-                subset = data_tczyx[[k - start_t]]
+        curr = data_tczyx
+
+        for level_index, level in enumerate(self.levels):
+            # downsample for levels > 0
+            if level_index > 0:
+                next_shape = (end_t - start_t,) + level.shape[1:]
+                curr = resize(curr, next_shape, order=0).astype(dtype)
```

Comment:
Is this an expensive operation? If so, could we skip this when the level exists (and overwrite is off) and calculate `curr` based on the current shape and level?

Reply:
Oh yeah this can be refactored